cpp/src/arrow/scalar.h

```diff
@@ -111,6 +111,12 @@ struct ARROW_EXPORT Scalar : public std::enable_shared_from_this<Scalar>,
  protected:
   Scalar(std::shared_ptr<DataType> type, bool is_valid)
       : type(std::move(type)), is_valid(is_valid) {}
+
+  // 16 bytes of scratch space to enable ArraySpan to be a view onto any
+  // Scalar- including binary scalars where we need to create a buffer
+  // that looks like two 32-bit or 64-bit offsets.
+  alignas(int64_t) std::array<uint8_t, sizeof(int64_t) * 2> scratch_space_{0};
```

Comment:
Why not `alignas(uint64_t) uint8_t scratch_space_[16];`?

I would leave out the 0 initialization as well. Leave it undefined for scalars that don't need the scratch buffers.

Reply:
This would also remove the need to call `.data()` above which was weird.

Reply:
Alright, I'll remove the zero init.

WRT std::array: I'm following the [linter](https://releases.llvm.org/13.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/modernize-avoid-c-arrays.html) which recommends avoiding C arrays in favor of `std::array`. If we would prefer to disable that lint check (and correspondingly prefer usage of C arrays), let's do that in a follow up

Reply:
@pitrou I think we should disable that `clang-tidy` rule. It's not a good fit for the kind of things we do in Arrow.

Reply:
@bkietz we don't fail CI build in `clang-tidy` failures though.